File: src/parser/enemy.ts
import * as _ from "lodash";
import { enemy_table } from "../data";
import { EnemyHandbook, EnemyData, EnemyTable } from "../data/enemy.i";
import { purge } from "../util";

export interface EnemyFlat {
  id: string;
  name: string;
  description: string;
  /** 序号 */
  index: string;
  /** 序号 */
  level: string;
  /** 评级endure,attack,defence,resistance */
  handbook: string[];
  enemyRace?: string;
  attackType: string;
  ability?: string;

  levels: EnemyAttribute[];
  refers: string[];
}
export interface EnemyAttribute {
  maxHp?: number;
  atk?: number;
  def?: number;
  magicResistance?: number;
  cost?: number;
  blockCnt?: number;
  moveSpeed?: number;
  attackSpeed?: number;
  baseAttackTime?: number;
  respawnTime?: number;
  hpRecoveryPerSec?: number;
  spRecoveryPerSec?: number;
  maxDeployCount?: number;
  massLevel?: number;
  baseForceLevel?: number;
  stunImmune?: boolean;
  silenceImmune?: boolean;
  lifePointReduce?: number;
  rangeRadius?: number;
}
export interface EnemyDetail extends EnemyAttribute {
  // id?: string;
  name?: string;
  description?: string;
}

export const parseEnemyData = (level: EnemyData, detail = true) => {
  return _.reduce(
    level,
    (r, v: any, n) => {
      if (n === "attributes") {
        _.forEach(v as EnemyTable[string][number]["enemyData"]["attributes"], (v, n) => {
          if (v.m_defined) r[n] = v.m_value;
        });
      } else if ((detail ? ["name", "description", "lifePointReduce", "rangeRadius"] : ["lifePointReduce", "rangeRadius"]).includes(n)) {
        if (v.m_defined) r[n] = v.m_value;
      }
      return r;
    },
    {}
  );
};

export const translateEnemy = (enemyHandbook: EnemyHandbook) => {
  const enemy = enemy_table[enemyHandbook.enemyId];
  const dst = {} as EnemyFlat;
  dst.id = enemyHandbook.enemyId;
  dst.name = enemyHandbook.name;
  dst.index = enemyHandbook.enemyIndex;
  dst.level = enemyHandbook.enemyLevel;
  dst.description = enemyHandbook.description;
  if (enemyHandbook.enemyRace) dst.enemyRace = enemyHandbook.enemyRace;
  dst.attackType = enemyHandbook.attackType;
  if (enemyHandbook.ability) dst.ability = enemyHandbook.ability;
  dst.handbook = [enemyHandbook.endure, enemyHandbook.attack, enemyHandbook.defence, enemyHandbook.resistance];
  if (enemy) {
    if (enemy.length < 1) {
      console.log(enemy);
      return;
    }
    dst.levels = enemy.map(v =>
      purge(parseEnemyData(v.enemyData, false), (v, n) => {
        if (n === "attackSpeed" && v === 100) return true;
        if (n === "massLevel" && v === 1) return true;
        if (n === "lifePointReduce" && v === 1) return true;
        return !v;
      })
    );
  }
  return dst;
};

File: src/module/Enemy.lua
local p = {}

local getArgs = require('Module:Arguments').getArgs
local EnemyData = mw.loadData('Module:Enemy/data')
local util = require('Module:Util')

function p.getEnemy(name)
  return EnemyData.Enemies[name]
end

function p.getEnemyFlat(name)
  if name == '持盾刀兵' then
    name = '机动盾兵'
  end
  local raw = p.getEnemy(name)
  local dst = {}
  dst['名称'] = name
  if not raw then
    return dst
  end
  dst['种族'] = raw.enemyRace or '其它'
  dst['级别'] = raw.level
  dst['编号'] = raw.index
  dst['id'] = raw.id
  dst['简介'] = raw.description
  dst['攻击类型'] = raw.attackType
  dst['能力'] = raw.ability or ''
  dst['出现关卡'] = raw.refers and util.join(raw.refers) or ''
  if raw.levels then
    if raw.levels[1] then
      dst['生命0'] = raw.levels[1].maxHp or 0
      dst['攻击0'] = raw.levels[1].atk or 0
      dst['防御0'] = raw.levels[1].def or 0
      dst['法术抗性0'] = raw.levels[1].magicResistance or 0
      dst['攻击速度0'] = raw.levels[1].baseAttackTime or 0
      dst['移动速度0'] = raw.levels[1].moveSpeed or 0
      dst['重量等级0'] = raw.levels[1].massLevel or 0
      dst['攻击距离0'] = raw.levels[1].rangeRadius or 0
    end
    if raw.levels[2] then
      dst['生命1'] = raw.levels[2].maxHp or raw.levels[1].maxHp or 0
      dst['攻击1'] = raw.levels[2].atk or raw.levels[1].atk or 0
      dst['防御1'] = raw.levels[2].def or raw.levels[1].def or 0
      dst['法术抗性1'] = raw.levels[2].magicResistance or raw.levels[1].magicResistance or 0
      dst['攻击速度1'] = raw.levels[2].baseAttackTime or raw.levels[1].baseAttackTime or 0
      dst['移动速度1'] = raw.levels[2].moveSpeed or raw.levels[1].moveSpeed or 0
      dst['重量等级1'] = raw.levels[2].massLevel or raw.levels[1].massLevel or 0
      dst['攻击距离1'] = raw.levels[2].rangeRadius or raw.levels[1].rangeRadius or 0
      dst['level1'] = 1
    end
  end

  -- 数组
  dst['属性'] = util.join(raw.handbook)
  return dst
end

function p.renderEnemy(name)
  local tab = p.getEnemyFlat(name)
  return mw.getCurrentFrame():expandTemplate {title = 'Enemy/template', args = tab}
end

function p.renderListEnemy(type)
  local rst = ''
  for _, enemyIndex in ipairs(EnemyData.EnemiesIndex) do
    local enemy = p.getEnemy(enemyIndex)
    if not type or type == enemy.level then
      rst = rst .. p.renderEnemy(enemy.name)
    end
  end
  return rst
end

function p.listEnemy(frame)
  local args = getArgs(frame)
  local type = args['type'] or args[1] or mw.title.getCurrentTitle().text
  local tab = p.renderListEnemy(type)
  return tab
end

function p.infobox(frame)
  local args = getArgs(frame)
  local name = args['name'] or args[1] or mw.title.getCurrentTitle().text
  local tab = p.getEnemyFlat(name)
  return mw.getCurrentFrame():expandTemplate {title = 'InfoboxEnemy/template', args = tab}
end

function p.enemy(frame)
  local args = getArgs(frame)
  local name = args['name'] or args[1] or mw.title.getCurrentTitle().text
  local tab = p.getEnemyFlat(name)
  return mw.getCurrentFrame():expandTemplate {title = 'Enemy/template', args = tab}
end

function p.link(frame)
  local args = getArgs(frame)
  local name = args['name'] or args[1] or mw.title.getCurrentTitle().text
  local tab = p.getEnemyFlat(name)
  return mw.getCurrentFrame():expandTemplate {title = 'LinkEnemy/template', args = tab}
end

function p.tt(frame)
  local args = getArgs(frame)
  local name = args['name'] or args[1] or mw.title.getCurrentTitle().text
  local tab = p.getEnemyFlat(name)
  return mw.getCurrentFrame():expandTemplate {title = 'tt/enemy/template', args = tab}
end

return p
